### ISSUE TITLE
[Proposal] Property access modifier to allow private setter access in object initializer (immutability)

### ISSUE THREAD

**Issue by nsarris** [dotnet/csharplang#1506]
Object immutability is currently implemented in c# with the use of constructors. However, this creates code that is both difficult to write (long constructors) and read. It also makes it impossible to use mapping expressions to construct immutable objects.

The proposition here is a new access modifier that explicitly dictates that a value can be set in an object initializer. It could for example be named "immutable" or "readonly" or "initonly" or "init". 

Example:

```csharp
public class Foo
{
	public string A { get; immutable set; }
	public string B { get; immutable set; }
}
```

Would allow this:

```csharp
var foo =  new Foo(){
	A = "a",
	B = "b"
}
```

but produce an error when trying to set the property elsewhere. For all intends and purposes it would be identical to the setter being public in an initializer and private anywhere else.

For anyone working with expressions or functional programming this has greate value. Besides the immutable classes themselves being cleaner and shorter, it maintains consistency when writing mapping expressions because currently you have do this (which is also not very intellisence friendly):

```csharp
query
.Select(x => new Foo(
	a: x.SomeProp, 
	b: x.SomeOtherProp, 
	c: x.AnotherProp))
```

vs this: (for muttable classes)

```csharp
query
.Select(x => new Foo {
	A = x.SomeProp, 
	B = x.SomeOtherProp, 
	C = x.AnotherProp})
```

Another issue with constructors is that you don't have an explicit map between the constructor parameter and the actual property it privately sets, so if you want to using dynamic mapping or cloning expressions with private setters you need to go by convention (name/sequence) or use attributes which means even more code or override the setter restrictions through reflection.

Finally this would make it very easy to refactor current code with immutability as it wouldn't need code changes other than the new accessor itself

**Comment by Joe4evr** [comment]
> currently you have do this (which is also not very intellisence friendly)

How is that snippet not IntelliSense-friendly? It gives you the parameter names and types the method expects. If you start typing the name, it'll suggest to complete the name for you.

**Comment by HaloFour** [comment]
How would this be implemented and enforced?  A property is just a method that can be invoked at some arbitrary time.  Neither the compiler nor the runtime have any concept of an "initialization" period after which the constructor is invoked.

**Comment by nsarris** [comment]
@Joe4evr 
In the context of an initializer intellisense offers only available properies. On constructors it offers everything in scope. You can filter besides being one more action it groups locals and parameters together. So Initializers are more intellisense friendly

@HaloFour 
A property is indeed method but accessors are syntactically enforced by the compiler and are translated to member properties at runtime (which can be overriden through reflection so they are not hard endorced). What I'm suggesting is already implemented in the compiler through anonymous types so it can be done. If you inspect anonymous types you will see that although they are initialized with initializers they have no setters at all. I haven't seen the produced IL but it could transpile it to constructor or field setter syntax during compilation.

**Comment by HaloFour** [comment]
@nsarris 

> What I'm suggesting is already implemented in the compiler through anonymous types so it can be done.

It's only possible with anonymous types because the compiler is in full control of their implementation and shape, so the compiler can choose to rewrite the initializer to constructor arguments.  You're always initializing all of the members.  But for arbitrary types that's not the case.  There's no requirement to initialize members and there is no inherent order of members from which to derive a constructor.

**Comment by nsarris** [comment]
@HaloFour 
I see so it's not possible like that. It could rewrite the class to add an appropriate constructor and call it but that would not work because it could be possible that it would need to rewrite overloads with the same kinds of paramters, so it won't work indeed

Would it then be possible to mark the whole class as immutable. An immutable object needs to have all it's members initialized, so they need default values. In constructor syntax that would be:
```csharp
public class Foo
{
    public string A { get; set; }
    public string B { get; set; }
    public string C { get; set; }

    public Foo(string a = "a", string b = null, string c ="bar")
    {
        A = a;
        B = b;
        C = c;
    }
}
```
You can imagine how this grows in large data objects

Would it instead be possible to write this
```csharp
public immutable class Foo
{
    public string A { get; } = "a";
    public string B { get; }
    public string C { get; } = "bar";
}
```
This would be rewritten as the initial declaration by the compiler. The constraint here would be that only value types and other immutable classes could be used as immutable members and no constructor can be explicitly declared.

**Comment by Joe4evr** [comment]
You're likely wanting the upcoming Record types (#39), where you can simplify that entirely to
```cs
public record class Foo(string A, string B, string C);
```

**Comment by HaloFour** [comment]
@Joe4evr 

You beat me to it.  I was trying to scan through the proposal to see if Records would support property initializers as well as constructor arguments but as far as I can tell they don't.

**Comment by nsarris** [comment]
I had noticed something like that, I thought they were called named tuples. Will they allow default values? If so it would be an acceptable compromise to use the constructor syntax but without having to specify all parameters every time when using named parameters. They initializer syntax with record types would be awesome though.

**Comment by Joe4evr** [comment]
> Will they allow default values?

I don't know. The current proposal doesn't mention it, and I don't think there's been real work on it yet due to prioritizing other big features (Nullable Reference Types, Default Implementations, etc.). However, that does mean they'll be a lot more open to outside input, so the best thing would just be to ask about it.

**Comment by MikeLeicester** [comment]
i would say that just extending the current` {get;} `syntax to cover constructors and initialisers would be more effecting than adding in a new access keyword

**Comment by HaloFour** [comment]
@MikeLeicester 

Initializers are just property setter calls that happen to be made sometime after the constructor is called.  It's impossible to distinguish them from normal settable properties.  Readonly properties are backed by `readonly` fields which the CLR forbids assigning to anywhere except from directly within the constructor.

**Comment by nsarris** [comment]
@HaloFour 

Actually that doesn't sound right. You can set readonly fields through reflection and IL injection. Also reflection has no relevant data to tell if a field was declared as readonly or not. Only expression compilation breaks when setting a readonly field which leads me to believe it's a purely syntactic constraint and not a CLR constraint.

@MikeLeicester 
I this is actually a better idea. How does the compiler handle setting properties with no getter in the constructor. I assume it rewrites them by replacing the property with the field itself. If the same principle is applied in the initializer, then it would only work if the compiler doesn't check the accessor (readonly/private) constraint after the rewrite.

**Comment by nsarris** [comment]
@HaloFour 
I didn't know that, thanks for the insight

@scalablecory 
From all the above it seems this is the more viable way to go but it would require default values for Records for uniformity as initializers don't require all properties set

**Comment by YairHalberstadt** [comment]
Closing as Championed as part of https://github.com/dotnet/csharplang/issues/39 and implemented in C# 9.0
